src/file-manager/fm-desktop-icon-view.c

```diff
@@ -708,11 +784,64 @@ action_change_background_callback (GtkAction *action,
 {
     g_assert (FM_DIRECTORY_VIEW (data));
 
-    caja_launch_application_from_command (gtk_widget_get_screen (GTK_WIDGET (data)),
-                                          _("Background"),
-                                          "mate-appearance-properties",
-                                          FALSE,
-                                          "--show-page=background", NULL);
+    if (GDK_IS_X11_DISPLAY (gdk_screen_get_display (gdk_screen_get_default())))
+        caja_launch_application_from_command (gtk_widget_get_screen (GTK_WIDGET (data)),
+                                              _("Background"),
+                                              "mate-appearance-properties",
+                                              FALSE,
+                                              "--show-page=background", NULL);
+    else
+    {
+        GtkWidget *dialog, *box;
+        char *filename;
+        GtkWidget *applybutton, *cancelbutton, *filebutton;
+        gint response;
+
+        filename = g_settings_get_string (background_settings, "picture-filename");
+
+        dialog = gtk_dialog_new ();
+        gtk_window_set_title (GTK_WINDOW (dialog), "Desktop Background Preferences");
+        gtk_window_set_transient_for (GTK_WINDOW (dialog), NULL);
+
+        applybutton = gtk_button_new_with_mnemonic (_("_Apply"));
+        cancelbutton = gtk_button_new_with_mnemonic (_("_Cancel"));
+
+        gtk_button_set_image (GTK_BUTTON (applybutton), gtk_image_new_from_icon_name ("gtk-ok", GTK_ICON_SIZE_BUTTON));
+        gtk_button_set_image (GTK_BUTTON (cancelbutton), gtk_image_new_from_icon_name ("gtk-cancel", GTK_ICON_SIZE_BUTTON));
+
+        gtk_button_set_use_underline (GTK_BUTTON (applybutton), TRUE);
+        gtk_button_set_use_underline (GTK_BUTTON (cancelbutton), TRUE);
+        gtk_widget_set_can_default (applybutton, TRUE);
+        gtk_dialog_add_action_widget (GTK_DIALOG (dialog), cancelbutton,GTK_RESPONSE_CANCEL);
+        gtk_dialog_add_action_widget (GTK_DIALOG (dialog), applybutton, GTK_RESPONSE_APPLY);
+        gtk_widget_show(applybutton);
+        gtk_widget_show(cancelbutton);
+
+        filebutton = gtk_file_chooser_button_new (_("Select a file"),
+                                        GTK_FILE_CHOOSER_ACTION_OPEN);
+
+        gtk_file_chooser_button_set_width_chars(GTK_FILE_CHOOSER_BUTTON (filebutton), 40);
+        gtk_file_chooser_set_filename (GTK_FILE_CHOOSER (filebutton), filename); 
+        gtk_file_chooser_button_set_title (GTK_FILE_CHOOSER_BUTTON (filebutton), "Select a File");
+
+        box = gtk_dialog_get_content_area(GTK_DIALOG(dialog));
+        gtk_box_pack_start (GTK_BOX(box), filebutton, FALSE, FALSE, 25);
+        gtk_widget_show(filebutton);
+
+        response = gtk_dialog_run (GTK_DIALOG(dialog));
+        if (response == GTK_RESPONSE_APPLY)
+        {
+            filename = gtk_file_chooser_get_filename (GTK_FILE_CHOOSER (filebutton));
```

Comment:
Memory leak

Reply:
Filename is a single variable whose value is set earlier in the function (for the existing value in the filechooser) and is reset if the user picks a new image from it. Whether they do or not, the code path cannot bypass 
`g_free (filename)`
at the end of this function, as it is after the` if` and` else` blocks